src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const stateCopy = {};
+
+  Object.assign(stateCopy, state);
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      for (const key in action.extraData) {
+        stateCopy[key] = action.extraData[key];
+      }
+
+      stateHistory.push(stateCopy);
+    } else if (action.type === 'removeProperties') {
+      for (const key in action.keysToRemove) {
+        delete stateCopy[key];
+      }
+
+      stateHistory.push(stateCopy);
+    } else if (action.type === 'clear') {
+      for (const key of Object.keys(state)) {
+        delete stateCopy[key];
+      }
+
+      stateHistory.push(stateCopy);
```

Comment:
As with previous cases, create a new copy of `stateCopy` before pushing it to `stateHistory` to correctly capture the state after each action.